src/node/mod.rs

```diff
@@ -0,0 +1,410 @@
+    stat(path.as_str()).is_ok()
+}
+
+#[cfg(target_os = "freebsd")]
+fn devname(dev: dev_t) -> Option<String> {
+    use std::os::raw::{c_char, c_int};
+
+    // Matching value of SPECNAMELEN in FreeBSD 13+
+    let mut dev_name = vec![0u8; 255];
+
+    let buf: *mut c_char = unsafe {
+        libc::devname_r(
+            dev,
+            libc::S_IFCHR, // Must be S_IFCHR or S_IFBLK
+            dev_name.as_mut_ptr() as *mut c_char,
+            dev_name.len() as c_int,
+        )
+    };
+
+    // Buffer was too small (weird issue with the size of buffer) or the device could not be named.
+    if buf.is_null() {
+        return None;
+    }
+
+    // SAFETY: The buffer written to by devname_r is guaranteed to be NUL terminated.
+    unsafe { dev_name.set_len(libc::strlen(buf)) };
+
+    Some(String::from_utf8(dev_name).expect("Returned device name is not valid utf8"))
+}
+
+/// Returns if the given device by major:minor pair is a drm device
+#[cfg(target_os = "freebsd")]
+pub fn is_device_drm(dev: dev_t) -> bool {
+    devname(dev).map_or(false, |dev_name| {
+        dev_name.starts_with("drm/")
+            || dev_name.starts_with("dri/card")
+            || dev_name.starts_with("dri/control")
+            || dev_name.starts_with("dri/renderD")
+    })
+}
+
+/// Returns if the given device by major:minor pair is a drm device
+#[cfg(not(any(target_os = "linux", target_os = "freebsd")))]
+pub fn is_device_drm(dev: dev_t) -> bool {
+    major(dev) == DRM_MAJOR
+}
+
+/// Returns the path of a specific type of node from the same DRM device as another path of the same node.
+pub fn path_to_type<P: AsRef<Path>>(path: P, ty: NodeType) -> io::Result<PathBuf> {
+    let stat = stat(path.as_ref()).map_err(Into::<io::Error>::into)?;
+    dev_path(stat.st_rdev, ty)
+}
+
+/// Returns the path of a specific type of node from the same DRM device as an existing DrmNode.
+pub fn node_path(node: &DrmNode, ty: NodeType) -> io::Result<PathBuf> {
+    dev_path(node.dev, ty)
+}
+
+/// Returns the path of a specific type of node from the DRM device described by major and minor device numbers.
+#[cfg(target_os = "linux")]
+pub fn dev_path(dev: dev_t, ty: NodeType) -> io::Result<PathBuf> {
+    use std::fs;
+    use std::io::ErrorKind;
+
+    if !is_device_drm(dev) {
+        return Err(io::Error::new(
+            ErrorKind::NotFound,
+            format!("{}:{} is no DRM device", major(dev), minor(dev)),
+        ));
+    }
+
+    let read = fs::read_dir(format!(
+        "/sys/dev/char/{}:{}/device/drm",
+        major(dev),
+        minor(dev)
+    ))?;
+
+    for entry in read.flatten() {
+        let name = entry.file_name();
+        let name = name.to_string_lossy();
+
+        // Only 1 primary, control and render node may exist simultaneously, so the
+        // first occurrence is good enough.
+        if name.starts_with(ty.minor_name_prefix()) {
+            let path = [r"/", "dev", "dri", &name].iter().collect::<PathBuf>();
```

Comment:
Same here, I expected this to use the generic function (with masking and shifting). Any comment here (and in  `is_device_drm()`) why these are special-cased for Linux and FreeBSD?

Reply:
The paths are different for freebsd and linux, so why wouldn't it be special-cased? A difference here is to be expected, no?

Reply:
But is there a difference between Linux and not(Linux, FreeBSD)?

Reply:
I see something weird on my machine which explains why you might not want to use IDs and bitmasks, but this logic checking inside `device/drm` instead:

```
❯ l /dev/dri/by-path/
total 0
drwxr-xr-x 2 root root 120 Sep  6 15:00 ./
drwxr-xr-x 3 root root 140 Sep  6 15:00 ../
lrwxrwxrwx 1 root root   8 Sep  6 15:00 pci-0000:23:00.0-card -> ../card1
lrwxrwxrwx 1 root root  13 Sep  6 15:00 pci-0000:23:00.0-render -> ../renderD128
lrwxrwxrwx 1 root root   8 Sep  6 15:00 pci-0000:4c:00.0-card -> ../card2
lrwxrwxrwx 1 root root  13 Sep  6 15:00 pci-0000:4c:00.0-render -> ../renderD129
```

```
❯ stat /dev/dri/* --format "%n: %R"
/dev/dri/by-path: 0
/dev/dri/card1: e201
/dev/dri/card2: e202
/dev/dri/renderD128: e280
/dev/dri/renderD129: e281
```

As you can see primary device `card1` has number `0` for the render device (if we mask the type bits out of it). Same for `card2` having number `1` in `renderD129`.